### PR TITLE
Add Exclusion Filters

### PR DESCRIPTION
**Pull request by glubsy**
* Allow users to specify files to filter out during scan by using regular expressions
* Set the default directory states according to regular expressions.
* File names and file paths are matched against regular expressions: only regexes with path separators are matches against the full path to the file, other regexes are matched against the filename only. This allows for more precise filtering.
* Two different implementations: one around a python list, and one around a python dict, but same features (I think the dict implementation might be a bit more efficient, so we use that by default now).
* Regexes can be combined (union) for possibly better performance, although my experiments didn't show meaningful results.
* Also added test suites.

### CODE REVIEW

**Comment by glubsy** [comment]
This GIF doesn't show it, but when mouse hovering over the red "denied" icon, there's a tooltip showing the error generated by the faulty regex.

This is based on the details dialog improvement branch. 
Also to note, this includes commits which fix rather serious bugs in the tabbed window feature (namely 3382bd5e5b5da1424b6e6d9b0eed256b503867fb, 26d18945b1c169b00f07d117456ef311c745556c, 584e9c92d9a28af6b9f627c40e78cbcc58bac455) 

Regarding the regex union feature, [here is a script](https://gist.github.com/glubsy/8ce410ee0304569d69e1f1f6d2b18c25) I wrote to roughly compare performance based on time. 
Unfortunately, the results vary a lot, sometimes unions are faster, sometimes they are not. There's probably a better way to test these but that's all I had time for.
In the end I doubt it matters very much.

File: qt/app.py [hunk 1]
```diff
@@ -223,6 +231,10 @@ def show_details(self):
     def showResultsWindow(self):
         if self.resultWindow is not None:
             if self.use_tabs:
+                if self.main_window.indexOfWidget(self.resultWindow) < 0:
+                    self.main_window.addTab(
+                        self.resultWindow, "Results", switch=True)
+                    return
```
Comment (arsenetar) [inline qt/app.py:237]: Seems like these changes are part of fixing something other than the title of the PR seems to state, I do see the commit for this particular item.  This is okay, but it makes it easier if changes for separate items are in separate PRs (okay with grouping some small misc fixes together in some cases).  This applies to potentially a few other lines and changes throughout, just commenting it here as the first location I looked at.

File: qt/tabbed_window.py [hunk 2]
```diff
@@ -142,6 +142,7 @@ def updateMenuBar(self, page_index=None):
             and not page_type == "IgnoreListDialog" else False)
         self.app.actionDirectoriesWindow.setEnabled(
             False if page_type == "DirectoriesDialog" else True)
+        # FIXME add ExcludeListDialog here too
```
Comment (arsenetar) [inline qt/tabbed_window.py]: Seems like this should be updated?

File: qt/tabbed_window.py [hunk 3]
```diff
@@ -223,14 +230,13 @@ def onTabCloseRequested(self, index):
             # menu or shortcut. But this is useless if we don't have a button
             # set up to make a close request anyway. This check could be removed.
             return
-        current_widget.close()
-        self.setTabVisible(index, False)
+        # current_widget.close()  # seems unnecessary
```
Comment (arsenetar) [inline qt/tabbed_window.py]: Seems like if these commented out lines are not needed we can just remove them.

File: tox.ini [hunk 4]
```diff
@@ -10,7 +10,7 @@ setenv =
     PYTHON="{envpython}"
 commands =
     make modules
-    py.test core hscommon
+    {posargs:py.test core hscommon}
```
Comment (arsenetar) [inline tox.ini]: Why the change here?

**Comment by glubsy** [comment]
Side note: just noticed [detwinner](https://neatdecisions.com/blog/2020/11/05/detwinner-linux-edition-0-4-0/) also added a similar regex-based filter for file paths. :wink: 

screen

**Review by arsenetar — CHANGES_REQUESTED** [review]
Few minor things here, also I am getting failures on the new directories and exclude tests.
```
FAILED core/tests/directories_test.py::TestExcludeList::test_exclude_refined - AssertionError: 6 != 5
FAILED core/tests/directories_test.py::TestExcludeList::test_get_state_returns_excluded_for_hidden_directories_and_files - AssertionError: 3 != 1
FAILED core/tests/directories_test.py::TestExcludeDict::test_exclude_refined - AssertionError: 6 != 5
FAILED core/tests/directories_test.py::TestExcludeDict::test_get_state_returns_excluded_for_hidden_directories_and_files - AssertionError: 3 != 1
FAILED core/tests/directories_test.py::TestExcludeListunion::test_exclude_refined - AssertionError: 6 != 5
FAILED core/tests/directories_test.py::TestExcludeListunion::test_get_state_returns_excluded_for_hidden_directories_and_files - AssertionError: 3 != 1
FAILED core/tests/directories_test.py::TestExcludeDictunion::test_exclude_refined - AssertionError: 6 != 5
FAILED core/tests/directories_test.py::TestExcludeDictunion::test_get_state_returns_excluded_for_hidden_directories_and_files - AssertionError: 3 != 1
FAILED core/tests/exclude_test.py::TestCaseCompiledList::test_compiled_files - AssertionError: 1 != 6
FAILED core/tests/exclude_test.py::TestCaseCompiledDict::test_compiled_files - AssertionError: 1 != 6
```

File: qt/preferences_dialog.py [hunk 5]
```diff
@@ -161,28 +161,31 @@ def _setupDisplayPage(self):
...
                                tr("Use bold font for references"))
-        gridlayout.addRow(self.reference_bold_font)
+        formlayout.addRow(self.reference_bold_font)
 
         self.result_table_ref_foreground_color = ColorPickerButton(self)
-        gridlayout.addRow(tr("Reference foreground color:"),
+        formlayout.addRow(tr("Reference foreground color:"),
                           self.result_table_ref_foreground_color)
         self.result_table_ref_background_color = ColorPickerButton(self)
         gridlayout.addRow(tr("Reference background color:"),
```
Comment (arsenetar) [inline qt/preferences_dialog.py]: Looks like this `gridlayout` should be `formlayout`.

File: qt/app.py (hunk 1 above)
Comment (glubsy) [reply qt/app.py:237]: Indeed. Ideally this would have been part of a separate PR, but these changes relied on changes from this particular branch (related to preference dialog window mostly), so I decided to keep there here. They are still in separate commits at least.

File: qt/preferences_dialog.py (hunk 5 above)
Comment (glubsy) [reply qt/preferences_dialog.py]: This was a merge issue which has been fixed in the other PR https://github.com/arsenetar/dupeguru/pull/733/commits/b16b6ecf4d5ca156682a74dfac32e2ace415f4e2 But I'll fix it here too if you prefer.

File: tox.ini (hunk 4 above)
Comment (glubsy) [reply tox.ini]: This allows positional arguments to call tox like so: `tox --skip-pkg-install -e py38 -- py.test "core/tests/exclude_test.py"`
It helps narrow down the test files to run, which speeds up the process when needed. It shouldn't affect anything other than giving the option.

File: qt/tabbed_window.py (hunk 2 above)
Comment (glubsy) [reply qt/tabbed_window.py]: This has already been fixed in a later commit.

**Comment by glubsy** [comment]
That's odd. I don't get any failure during the tests. The Continuous Integration system doesn't seem to produce these either. Could it be that these occur on Windows only? 

Edit: nevermind, I forgot you had set up a Windows machine for the CI too. Still no error there. 
I'll run these tests in Win10 and see if I can reproduce.

**Comment by arsenetar** [comment]
@glubsy the tox environment for windows does not actually run the tests... so the CI will not catch that.  I should change it to actually run the test, all it does right now is run build and packaging.

File: tox.ini (hunk 4 above)
Comment (glubsy) [reply tox.ini]: This is the source documentation on the matter: https://tox.readthedocs.io/en/latest/config.html#interactive-shell-substitution

Comment (arsenetar) [reply tox.ini]: I'm fine with that, I am probably going to make some change to the CI to streamline this configuration anyhow.

**Comment by arsenetar** [comment]
@glubsy I updated the CI to actually run the tests on windows, which should help show these failures, we do have a minor conflict with my changes in the tox.ini now as I brought in another change there.  I took a brief look at the tests that were failing and it seems to be that some of these have path separators in the regex which might explain why they are not filtering out when they should be (windows paths using different separator).

**Comment by glubsy** [comment]
Sorry for the late reply. I could indeed reproduce the issue on Windows 7 and already fixed half of it. I'll push the rest of the changes in a bit (it's 5AM here, I have to take a break).

Thanks for updating the CI!

**Review by arsenetar — APPROVED** [review]
